# Going further 2: numbers in the real world
After Lecture 2 (Variables & Types). Same `int`, `float` and math symbols, but with real stories:
a YouTube counter that broke, a 0.34 second mistake that mattered, and how far light travels.

> 🚀 **Going further** is optional: not in the homework, not on the exam. It is for students who
> finished the lecture notebooks and want to see more. Everything here runs in Colab, nothing to install.

## 1. Python's whole numbers never run out
In many languages a whole number lives in a fixed box of 32 bits. The biggest number that fits is
`2 ** 31 - 1`. Python's `int` grows as needed, so it has no such limit.

In [1]:
biggest_32_bit = 2 ** 31 - 1
print(biggest_32_bit)
print(2 ** 100)

2147483647
1267650600228229401496703205376


**Real world:** in December 2014 the video *Gangnam Style* passed 2,147,483,647 views, exactly that biggest
32-bit number. YouTube had to switch its view counter to a bigger box (64 bits). In Python this would
never have happened.

## 2. Decimals are (almost) never exact
🔮 **Predict first:** what is `0.1 + 0.2`?

In [2]:
print(0.1 + 0.2)
print(0.1 + 0.2 == 0.3)

0.30000000000000004
False


The computer stores numbers in **binary** (0s and 1s). Like 1/3 = 0.3333... never ends in decimal,
0.1 never ends in binary, so it is stored slightly wrong. Tiny, but it adds up.

**Real world (1991):** a Patriot missile battery counted time in tenths of a second. Each tenth was stored
with an error of about 0.000000095 seconds. The system had been running for 100 hours. How big was the error?

In [3]:
error_per_tick = 0.000000095   # seconds lost every tenth of a second
ticks_per_second = 10
hours = 100
ticks = ticks_per_second * 60 * 60 * hours
total_error = ticks * error_per_tick
print(ticks)
print(total_error)

3600000
0.342


About **0.34 seconds**. A Scud missile flies about 1,676 meters every second. How far off was the radar?

In [4]:
scud_speed = 1676   # meters per second
print(total_error * scud_speed)

573.192


More than half a kilometer, so the radar looked in the wrong place and the missile was not intercepted.
This is why **banks count money in whole agorot (an `int`), not in shekels with decimals**.

In [5]:
# Money the safe way: whole agorot
price_agorot = 1290          # 12.90 NIS
total_agorot = 3 * price_agorot
print(total_agorot // 100)   # shekels
print(total_agorot % 100)    # agorot

38
70


## 3. Huge and tiny numbers: scientific notation
`1.496e11` means 1.496 × 10¹¹. This is how scientists write numbers that are too long to read.

In [6]:
sun_distance = 1.496e11   # meters from the Sun to Earth
light_speed = 3e8          # meters per second
seconds = sun_distance / light_speed
print(seconds)
print(int(seconds) // 60)   # whole minutes
print(int(seconds) % 60)    # leftover seconds

498.6666666666667
8
18


The sunlight you see left the Sun **about 8 minutes and 18 seconds ago**. Our old friends `//` and `%`, used by astronomers.
(`int()` turns the decimal into a whole number. You will meet it properly in Lecture 3.)

## 4. Money grows with `**`
1,000 NIS in a savings plan that gives 5% a year. After 1 year: `1000 * 1.05`. After 10 years: `1000 * 1.05 ** 10`.

In [7]:
savings = 1000
rate = 1.05
print(savings * rate ** 10)
print(savings * rate ** 30)

1628.894626777442
4321.942375150667


Money roughly doubles every 14 years at 5%. Bankers use the **rule of 72**: years to double ≈ 72 / rate in percent.

In [8]:
print(72 / 5)

14.4


## Your turn
1. How many seconds old are you? (years × 365 × 24 × 60 × 60). Is it bigger than `2 ** 31 - 1`?
2. Light from the Moon: the Moon is `3.84e8` meters away. How many seconds does its light take?
3. The rule of 72 at 8% a year: how many years to double? Check it with `**`.

In [9]:
# 1.
age_years = 20
print(0)

0


In [10]:
# 2.
print(0)

0


In [11]:
# 3.
print(0)

0
